test/spark/run-client-test.sh

```diff
@@ -1,44 +1,121 @@
+    if [[ ${days_ago} -gt -1 ]]
+    then
+      run_lakectl fs rm "lakefs://${repo}/${branch_name}/file"
+      epoch_commit_date_in_seconds=$(( ${current_epoch_in_seconds} - ${day_in_seconds} * ${days_ago} ))
+      run_lakectl commit "lakefs://${repo}/${branch_name}" --allow-empty-message --epoch-time-seconds ${epoch_commit_date_in_seconds}
+    else   # This means that the branch should be deleted
+      run_lakectl branch delete "lakefs://${repo}/${branch_name}" -y
+    fi
+  done
+}
 
+last_commit_ref() {
+  local repo=$1
+  local branch=$2
+  run_lakectl log lakefs://${repo}/${branch} --amount 1 | grep "ID: " | awk '{ print $2 }'
+}
 
-# Validate Garbage Collection
-for f in \
-    lakefs://test-data/09d9016a12777fe/commits/commit-four \
-    lakefs://test-data/18c9a1fe4695109/commits/commit-two \
-    lakefs://test-data/efb14f67a2a3c38/commits/commit-one \
-    lakefs://test-data/dede737a85e27f4/commits/commit-five
-do
-    if docker-compose exec -T lakefs lakectl fs cat $f> /dev/null 2>&1 ; then
-      echo "Expected $f to be removed by garbage collection"
-      exit 1
+validate_gc_job() {
+  local test_case=$1
+  local repo=$2
+  local existing_ref=$3
+  local file_should_be_deleted=$(echo ${test_case} | jq -r '.file_deleted')
+  if run_lakectl fs cat lakefs://${repo}/${existing_ref}/file > /dev/null 2>&1 && ${file_should_be_deleted} ; then
+    echo "Expected the file to be removed by the garbage collector but it has remained in the repository"
+    exit 1
+  elif ! run_lakectl fs cat lakefs://${repo}/${existing_ref}/file > /dev/null && ! ${file_should_be_deleted} ; then
+    echo "Expected the file to remain in the repository but it was removed by the garbage collector"
+    exit 1
+  fi
+  echo ${test_case} | jq -c '.branches []' | while read branch_props; do
+    local days_ago=$(echo ${branch_props} | jq -r '.delete_commit_days_ago')
+    if [[ ${days_ago} -gt -1 ]]; then
+      local branch_name=$(echo ${branch_props} | jq -r '.branch_name')
+      for location in \
+        lakefs://${repo}/${branch_name}/not_deleted_file1 \
+        lakefs://${repo}/${branch_name}/not_deleted_file2 \
+        lakefs://${repo}/${branch_name}/not_deleted_file3
+      do
+        if ! run_lakectl fs cat ${location} > /dev/null; then
+          echo "expected ${location} to exist"
+          exit 1
+        fi
+      done
     fi
-done
+  done
+}
+
+clean_main() {
+  local repo=$1
+  run_lakectl fs rm "lakefs://${repo}/main/not_deleted_file1"
+  run_lakectl fs rm "lakefs://${repo}/main/not_deleted_file2"
+  run_lakectl fs rm "lakefs://${repo}/main/not_deleted_file3"
+  run_lakectl commit "lakefs://${repo}/main" -m "delete the undeleted files"
+}
+
+#################################
+######## Tests Execution ########
+#################################
 
-for f in \
-    lakefs://test-data/697297df0c01d17/commits/commit-ten \
-    lakefs://test-data/651a755f7f5b1a4/commits/commit-three
+day_in_seconds=86400
```

Comment:
This is not always true, there are leap seconds and not all days have 86_400 seconds.  That will might the test occasionally flaky -- one day every few years.

GNU date lets you say things like
```sh
date +%s --date '2 days ago'
```
to epoch time for the day before yesterday.

Reply:
The problem with that is that it really works with GNU only. So if we want to run it locally that might be an issue (for all of us Mac fan girls/boys).
I don't think that a leap second is a big issue in our context, first, because the occasional flakiness is once every few years, and second, even if a leap second occurs it will probably not fail the tests because of the time it takes to initialize the GC process and send network calls (which will take more than a second) which happens after we delete and commit.